collector/receiver/otelarrowreceiver/internal/arrow/arrow.go

```diff
@@ -370,163 +369,338 @@ func (r *Receiver) anyStream(serverStream anyStreamServer, method string) (retEr
+func (r *Receiver) recvOne(streamCtx context.Context, serverStream anyStreamServer, hrcv *headerReceiver, pendingCh chan<- batchResp, method string, ac arrowRecord.ConsumerAPI) (retErr error) {
 
 	// Receive a batch corresponding with one ptrace.Traces, pmetric.Metrics,
 	// or plog.Logs item.
 	req, err := serverStream.Recv()
 
+	// inflightCtx is carried through into consumeAndProcess on the success path.
+	inflightCtx, flight := r.newInFlightData(streamCtx, method, req.GetBatchId(), pendingCh)
+	defer flight.recvDone(inflightCtx, &retErr)
+
+	// this span is a child of the inflight, covering the Arrow decode, Auth, etc.
+	_, span := r.tracer.Start(inflightCtx, "otel_arrow_stream_recv")
+	defer span.End()
+
 	if err != nil {
 		if errors.Is(err, io.EOF) {
 			return status.Error(codes.Canceled, "client stream shutdown")
 		} else if errors.Is(err, context.Canceled) {
 			return status.Error(codes.Canceled, "server stream shutdown")
 		}
+		// Note: err is directly from gRPC, should already have status.
 		return err
 	}
 
 	// Check for optional headers and set the incoming context.
-	thisCtx, authHdrs, err := hrcv.combineHeaders(ctx, req.GetHeaders())
+	inflightCtx, authHdrs, err := hrcv.combineHeaders(inflightCtx, req.GetHeaders())
 	if err != nil {
 		// Failing to parse the incoming headers breaks the stream.
-		return fmt.Errorf("arrow metadata error: %w", err)
+		return status.Errorf(codes.Internal, "arrow metadata error: %v", err)
 	}
-	var prevAcquiredBytes int
-	uncompSizeStr, sizeHeaderFound := authHdrs["otlp-pdata-size"]
-	if !sizeHeaderFound || len(uncompSizeStr) == 0 {
+	var prevAcquiredBytes int64
+	uncompSizeHeaderStr, uncompSizeHeaderFound := authHdrs["otlp-pdata-size"]
+	if !uncompSizeHeaderFound || len(uncompSizeHeaderStr) == 0 {
 		// This is a compressed size so make sure to acquire the difference when request is decompressed.
-		prevAcquiredBytes = proto.Size(req)
+		prevAcquiredBytes = int64(proto.Size(req))
 	} else {
-		prevAcquiredBytes, err = strconv.Atoi(uncompSizeStr[0])
+		prevAcquiredBytes, err = strconv.ParseInt(uncompSizeHeaderStr[0], 10, 64)
 		if err != nil {
-			return fmt.Errorf("failed to convert string to request size: %w", err)
+			return status.Errorf(codes.Internal, "failed to convert string to request size: %v", err)
 		}
 	}
-	// bounded queue to memory limit based on incoming uncompressed request size and waiters.
-	// Acquire will fail immediately if there are too many waiters,
-	// or will otherwise block until timeout or enough memory becomes available.
-	err = r.boundedQueue.Acquire(ctx, int64(prevAcquiredBytes))
-	if err != nil {
-		return fmt.Errorf("breaking stream: %w", err)
+	// Authorize the request, if configured, prior to acquiring resources.
+	if r.authServer != nil {
+		var authErr error
+		inflightCtx, authErr = r.authServer.Authenticate(inflightCtx, authHdrs)
+		if authErr != nil {
+			flight.replyToCaller(status.Error(codes.Unauthenticated, authErr.Error()))
+			return nil
+		}
 	}
```

Comment:
Can't we do this before the `prevAcquiredBytes` logic?

Reply:
I understand wanting to authenticate before doing any work, but some amount of memory has already been used by this point.  Is it worth it to use additional resources to authenticate before memory limiting?

Reply:
I think we can do authentication before resource limiting, and yes the call to Auth can happen before prevAcquiredBytes is calculated.

Do you agree with this order? If the Auth fails afterwards, the acquired resource will be released immediately, so it seems to not matter very much.